File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
    const isBST = function (curr, low, high) {
        if (!curr) return true;

        if ((low !== null && low >= curr.val) || (high !== null && high <= curr.val)) return false;

        return isBST(curr.left, low, curr.val) && isBST(curr.right, curr.val, high);
    }

    return isBST(root, null, null);
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isBST(curr, low, high):
            if curr == None:
                return True

            if (low != None and low >= curr.val) or (high != None and high <= curr.val):
                return False

            return isBST(curr.left, low, curr.val) and isBST(curr.right, curr.val, high)

        return isBST(root, None, None)
